ambry-replication/src/main/java/com.github.ambry.replication/ReplicationManager.java

```diff
@@ -115,8 +127,89 @@ public void start() throws ReplicationException {
         this.scheduler.scheduleAtFixedRate(persistor, replicationConfig.replicationTokenFlushDelaySeconds,
             replicationConfig.replicationTokenFlushIntervalSeconds, TimeUnit.SECONDS);
       }
+      started = true;
     } catch (IOException e) {
       logger.error("IO error while starting replication", e);
+    } finally {
+      startupLatch.countDown();
+    }
+  }
+
+  /**
+   * {@inheritDoc}
+   * Note that, this method should be thread-safe because multiple threads (from different cluster change handlers) may
+   * concurrently update remote replica infos.
+   */
+  @Override
+  public void onReplicaAddedOrRemoved(List<ReplicaId> addedReplicas, List<ReplicaId> removedReplicas) {
```

Comment:
Looks like this method will be called either for replicas being added or removed. We should use a boolean then (e.g, isRemove), instead of two lists.

Reply:
I see your point but the method is invoked by InstanceConfig change and there could be both added and removed replica within a single InstanceConfig change. We pass in two lists (can be empty) and let replication manager handle both cases at the same time.